### ISSUE TITLE
Please don't lock the tsconfig.json file

### ISSUE THREAD

**Issue by claycephas** [Microsoft/vscode#1032]
I am attempting to use a separate git client and it keeps complaining about the tsconfig.json file being locked. If I close the folder from Visual Studio Code, the file is unlocked but also deleted.

It looks like Visual Studio Code is somehow locking the tsconfig.json file.

Expected:
Allow other applications to read and write the tsconfig.json file even if the folder containing that file is open in Visual Studio Code.

Environment:
Visual Studio Code:
![image](https://cloud.githubusercontent.com/assets/469582/11603997/6f6b1c42-9aa4-11e5-8155-23ba85831bde.png)

OS: Windows 7

**Comment by dbaeumer** [comment]
@claycephas  we don't do anything special with the tsconfig.json file and I wasn't able to reproduce this. Can you please use the handle tool from sysinternals (https://technet.microsoft.com/en-us/sysinternals/bb896655.aspx) to see who is locking the file in your case?

**Comment by Clivar** [comment]
I have the same isue. I dont explicitly open the file but I have a task that modifies it for me. However even if that task is killed I still have to close Visual studio code to unlock the file.

**Comment by tonyeung** [comment]
+1

**Comment by Clivar** [comment]
It happened again... code.exe is locking the file according to handle

**Comment by dbaeumer** [comment]
@Clivar do you have any additional steps. I tried various scenarios (having tsconfig.json open, changing it, ....) but I couldn't make code holding a handle to it. 

I suspect that the tsserver is holding the handle but before moving it to them I want to make sure that this is the case. When it happens the next time can you provide the command line of the code.exe process holding the lock. That would help ti understand which part of the code is the culprit here since code is multi process  and one VSCode instance start ~6 code.exe processes.

**Comment by tonyeung** [comment]
fwiw, try installing gitex and see if that has to do with anything. The last time I remember it happening was when I checkout/merge branches. Intellisense went crazy and I finally tracked it down to the tsconfig.json being missing. Happened a few more times with the tsconfig.json either being locked or gone all together.

**Comment by Clivar** [comment]
@tonyeung Exactly, I have it when pulling or merging with git. File went missing or was empty sometimes aswell. Happens quite a bit. Closing and reopening code has always fixed the access denied issue so far

**Comment by tonyeung** [comment]
@Clivar  how do you interact with git (git bash or gitex or sourcetree)?

**Comment by Clivar** [comment]
git bash. I do however commit from inside vs code

**Comment by tonyeung** [comment]
@Clivar  thanks. So to clarify, you commit within vscode, but do execute other git functions via git bash right?

**Comment by claycephas** [comment]
I have the same issue of `tsconfig.json` file is locked and after exiting VS code the `tsconfig.json` file is deleted.

I use Git Extensions on Windows for all git commands (commit, push, pull).

    C:\Apps>handle tsconfig.json

    Handle v4.0
    Copyright (C) 1997-2014 Mark Russinovich
    Sysinternals - www.sysinternals.com

    code.exe           pid: 16764  type: File           260: C:\Source\Modus\client\tsconfig.json
    code.exe           pid: 12804  type: File           260: C:\Source\VSS.UI.Utilities\tsconfig.json


PID 12804:

    "C:\Program Files (x86)\Microsoft VS Code\Code.exe" "c:\Program Files (x86)\Microsoft VS Code\resources\app\extensions\typescript\out\utils\electronForkStart" "c:\Program Files (x86)\Microsoft VS Code\resources\app\extensions\typescript\server\typescript\lib\tsserver.js"


PID 16764:

    "C:\Program Files (x86)\Microsoft VS Code\Code.exe" "c:\Program Files (x86)\Microsoft VS Code\resources\app\extensions\typescript\out\utils\electronForkStart" "c:\Program Files (x86)\Microsoft VS Code\resources\app\extensions\typescript\server\typescript\lib\tsserver.js"

**Comment by Clivar** [comment]
@tonyeung Exactly, Pull and Push sometimes from UI aswell but most of the time I use GIT Bash for repo actions

**Comment by claycephas** [comment]
Was able to reproduce the tsconfig.json file lock again. Here's the output to `handle.exe`:

    Handle v4.0
    Copyright (C) 1997-2014 Mark Russinovich
    Sysinternals - www.sysinternals.com
    
    code.exe pid: 16136 type: File 24C: C:\VSS.UI.Utilities\VSS
    .UI.Utilities\tsconfig.json


Process Explorer shows this is the command line for PID 16136:

    "C:\Program Files (x86)\Microsoft VS Code\code.exe" "c:\Program Files (x86)\Microsoft VS Code\resources\app\extensions\typescript\out\utils\electronForkStart" "c:\Program Files (x86)\Microsoft VS Code\resources\app\extensions\typescript\server\typescript\lib\tsserver.js"

After exiting Visual Studio Code the tsconfig.json file was deleted.

**Comment by tonyeung** [comment]
I think the deletion of the tsconfig.json file makes this a pretty critical bug. 

The first time it happened it took me a while to figure out what happened. The worse part is that the built in git client didn't pick up the change for a while, so I didn't notice anything missing. If for some reason someone was working without source control it would be a nightmare to figure out why intellisense doesn't work anymore.

**Comment by glen-84** [comment]
I've seen the same thing with other files (it was a TypeScript file this time):

```
$ git rebase upstream/master
First, rewinding head to replay your work on top of it...
error: unable to create file gulp-typescript/gulp-typescript.d.ts (Permission denied)
Fast-forwarded master to upstream/master.

$ git rebase upstream/master
gulp-typescript/gulp-typescript.d.ts: Permission denied
Current branch master is up to date.

$ git rebase upstream/master
Cannot rebase: You have unstaged changes.
Please commit or stash them.

$ git status
On branch master
Your branch is ahead of 'origin/master' by 627 commits.
  (use "git push" to publish your local commits)
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        deleted:    gulp-typescript/gulp-typescript.d.ts

no changes added to commit (use "git add" and/or "git commit -a")
```

The file was open inside VS Code.

Windows 10.

**Comment by dbaeumer** [comment]
@claycephas Thanks a lot for the information. As suspected it is the tsserver that locks the tsconfig.json file. I will open a bug against the TypeScript team. I have one additional question: are you seeing this with the latest VSCode as well (e.g. 0.10.6). This version ships with a new version of the tsserver.

**Comment by dbaeumer** [comment]
Open https://github.com/Microsoft/TypeScript/issues/6337 in TS land

**Comment by tonyeung** [comment]
@dbaeumer  i just had an issue with the tsconfig.json file again last night.. or was it early this morning.. I'm pretty sure I'm on the latest vs code.

**Comment by Penryn** [comment]
I'm having a weird problem with tsconfig.json when VSCode is open. If I do a `git checkout someBranch` tsconfig.json sometimes gets deleted or locked and I have to issue a `git reset --hard HEAD` despite having just switched branches. tsconfig.json exists in both branches.

All git commands are issued by commandline (ie: not through VSCode).

I think this is caused by VSCode locking the tsconfig file so git can't overwrite it or something, and it eventually ends up deleted. If VSCode is closed, this doesn't happen. However, no git errors ever occur and I don't notice until my tsc commands fail or I do a `git status`. Pretty fishy.

**Comment by tonyeung** [comment]
@Penryn  its an issue with tsc, @dbaeumer  already opened a ticket with the typescript team.

**Comment by Penryn** [comment]
Ah, thank you for the update. I understand now.

**Comment by pocesar** [comment]
vscode keeps deleting my tsconfig.json whenever I checkout a new git branch

**Comment by dbaeumer** [comment]
@pocesar: could it be that git deletes it. We don't have any logic in regards to git operations. (e.g. what happens if you checkout the branch without having VSCode open)

**Comment by pocesar** [comment]
@dbaeumer nope, tsconfig.json is on all of my branches

**Comment by dbaeumer** [comment]
@pocesar: is your repository public so that I can try to reproduce.

**Comment by claycephas** [comment]
Another repro with VS Code version `0.10.6`.

![image](https://cloud.githubusercontent.com/assets/469582/12495191/93a9aa7a-c04b-11e5-81b3-4bbed3acd0e0.png)

Output from `handle.exe`:

    C:\Users\canders2>c:\apps\Handle.exe tsconfig.json

    Handle v4.0
    Copyright (C) 1997-2014 Mark Russinovich
    Sysinternals - www.sysinternals.com

    code.exe           pid: 1520   type: File           260: C:\Source\Modus\client\tsconfig.json


Output from `procexp.exe`

    Path:
    C:\Program Files (x86)\Microsoft VS Code\code.exe
    
    Command line:
    "C:\Program Files (x86)\Microsoft VS Code\Code.exe" "c:\Program Files (x86)\Microsoft VS Code\resources\app\extensions\typescript\out\utils\electronForkStart" "c:\Program Files (x86)\Microsoft VS Code\resources\app\extensions\typescript\server\typescript\lib\tsserver.js"
    
    Current Directory:
    C:\Program Files (x86)\Microsoft VS Code\
    
    PARENT:
    PID 15536
    
    PARENT Path:
    C:\Program Files (x86)\Microsoft VS Code\code.exe
    
    PARENT Command line:
    "C:\Program Files (x86)\Microsoft VS Code\Code.exe" "c:\Program Files (x86)\Microsoft VS Code\resources\app\out\bootstrap" --type=pluginHost
    
    PARENT Current directory:
    C:\Program Files (x86)\Microsoft VS Code\
    
After killing PID 1520: `client\tsconfig.json` file is deleted.
![image](https://cloud.githubusercontent.com/assets/469582/12495482/1ea58080-c04d-11e5-9ff1-ee7138a7392d.png)

**Comment by dbaeumer** [comment]
@claycephas did you run a git operation when that happened (e.g. switching branches). I think the deletion of the tsconfig.json file happens as follows: 

- some one tries to delete tsconfig.json. However since it is locked it fails, however to my knowledge Windows queues the delete and executes it when the lock is gone.
- when VSCode ends the file will be deleted.

**Comment by pocesar** [comment]
@dbaeumer it's private, but like was demonstrated by other people in here, it's fairly easy to reproduce. I opened the Devtools inside vscode, and caught this error

![vscode](https://cloud.githubusercontent.com/assets/461084/12537046/cc4a2d7a-c29c-11e5-915a-98ccd511b530.jpg)

intellisense for Typescript stops completly, even though my build task is running

**Comment by claycephas** [comment]
@dbaeumer, the command I was running was `git checkout` to switch branches.

**Comment by bpasero** [comment]
@dbaeumer I can reproduce on Windows 10. Here are some things I see that are weird:
* it reproduces only once I open a TS file, so I speculate TS-Server locks the file
* it does not reproduce in 0.10.7-dev
* it does also not reproduce for me when I run 0.10.6 "out of dev"

I am puzzled what kind of difference it could make to run 0.10.6 built or "out of dev" with regards to file locking.

**Comment by dbaeumer** [comment]
As puzzles as I am :-). As said today I was not able to reproduce this in the upcoming 0.10.7-dev build with the identical tsserver bits. When testing under 0.10.6 I noticed that the first file the tsserver reads get locked. But I have no explanation for it.

@bpasero you mentioned today that you disabled asar support. Could it be that asar support is still on in 0.10.6 when using code.exe as a node environment and in 0.10.7 it is off. That could be an explanation for the behavior.

**Comment by bpasero** [comment]
@dbaeumer I reverted my change for noasar and it did not make a difference. Unfortunately I am not able to easily switch back to the Electron we shipped in 0.10.6 because of all the API changes in Electron that I adopted, otherwise we could nail this down more easily.

Btw I also once commented out all file watching code in TS-Server and could still reproduce the file locking issue, so it seems to be unrelated.

**Comment by dbaeumer** [comment]
@bpasero yes, it is unrelated to file watching. I was able to reproduce on 0.10.6 with a simple fs.readFileSync but not on 0.10.7.

**Comment by dbaeumer** [comment]
I will close this as fixed for 0.10.7 since all test cases we have that fail on 0.10.6 pass on 0.10.7 although we don't have the final answer why. 

@bpasero can you please marked as verified.

**Comment by pke** [comment]
This problem is in all Electron based editors. Atom has the same probs. So the real problem lays in the way Electron (Nodejs?) opens files on Windows. It should open them with no lock of course.
